### PR TITLE
Updating: Dependencies, test suite, bundling, and documenation

### PR DESCRIPTION
**Pull request by rschristian**
## Description

(Sorry in advance for this large PR, but many of the changes build off of each other. Certainly tell me to break it up if this is too much)

This PR updates many parts of this template that either were outdated, didn't match other templates, or simply needed cleaning. Look below for the full change log

### Changelog
* [x]  Update all dependencies to newest(-ish)
* [x]  Move dependencies into right categories -- ESLint & types packages don't belong under `dependencies`
* [x]  Refactor tests to use `enzyme` instead of `identity-obj-proxy`
* [x]  Add `microbundle` to do NPM-ready bundling instead of heavily editing the `preact-cli` to get that done
* [x]  Drop Sass - It was not being used and as not everyone wants SASS, plain CSS is a better default
* [x]  Switch template from EJS to HTML. May as well now that the CLI provides an HTML template to use
* [x]  Update documentation where appropriate
* [x]  Move demo items out of `src` and into a dedicated directory -- cleans up `src` and separates the concerns better IMO
* [x]  Move tests out of `src` and into a dedicated directory -- matches default template

### CODE REVIEW

**Comment by prateekbh** [comment]
This is dope @RyanChristian4427 🎉

File: template/tsconfig.test.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "extends": "tsconfig.json",
+  "include": ["src/**/*.tsx", "src/**/*.ts"]
+}
```
Comment (RyanChristian4427) [inline template/tsconfig.test.json:4]: This is a bit awkward but I figured the best way to go about the issue. 

To ensure the tests directory isn't bundled into our output, I exclude everything but `components` from the main `tsconfig.json`. However, my IDE (Webstorm), and I'm sure others, will raise errors in the non-`components` files as now there are `.ts` files that are not covered by a `tsconfig.json`. This is reasonable, the `--jsx` flag is needed. However, this PR uses `jest-preset-preact` to do our ts->js transformations, and it just uses Babel. It doesn't care about our config one bit. 

So this config is essentially just here to silence any warnings users might have. They can tell their IDEs to use this config, which covers all files, while the main one will just cover the files we actually want to output.

There might be a better way to go about this, not 100% sure.

**Comment by camsloanftc** [comment]
Just coming to say I was having some significant build time issues on the main branch, so I ended up pulling down a lot of what is in this PR and it works a lot better.

So I guess I just want to ask about the status of this PR to see if it makes sense to merge into master, or if it's blocked by anything. Would love it if other devs had a smoother initial onboarding experience than I did. Cheers.

**Comment by rschristian** [comment]
Glad this could be of some use. Did you happen to run into any issues with this update or functionality you thought I left out? Feedback is always welcome.

As for the status, I'll try to get an update here.

**Comment by camsloanftc** [comment]
Thanks @rschristian. I didn't have much success with the microbundle dist script. I might just be misconfiguring things, or it could be my lack of experience with widgets in general.
I was not able to get a version that builds with all of my dependencies bundled in and working properly in consuming projects (locally), so I kind of just went back to using the standard preact build script for now and have put that off for another day when I actually need to ship this widget up to a CDN for consuming in the browser.

I guess related to this, you had removed the `preact.config.js` file in this PR, and so my js bundles were being uniquely named each time, so when I went to consume in another project for testing (which I needed to do since it is a widget), I would need to know the specific bundle file hashed name. I ended up just putting in my own preact.config.js file (see below) to set it back to a consistent name of `bundle.js`.

```js
webpack(config, env) {
  config.output.filename = '[name].js'
}
```

But this might also be temporary, since it is only to do with local development still. I can see the value in the hashed names when needing to cache bust from client sites. I may still be too early in my development to give feedback on this part.

**Comment by rschristian** [comment]
> I was not able to get a version that builds with all of my dependencies bundled in and working properly in consuming projects (locally)

Hm, my first guess would be to check the the `package.json`'s `source` along with the `tsconfig.json`. Happy to troubleshoot over on the Slack (which I see you've joined). 

> I guess related to this, you had removed the preact.config.js file in this PR, and so my js bundles were being uniquely named each time, so when I went to consume in another project for testing (which I needed to do since it is a widget), I would need to know the specific bundle file hashed name.

Yeah, that was on purpose, though I suppose I could revert. The general idea was that you'd test your widget with Preact-CLI in it's own repo, rather than importing it somewhere else to test it. Preact-CLI is here entirely for application purposes, hence the hashed name. 

> I can see the value in the hashed names when needing to cache bust from client sites.

So here's the other part, you could potentially turn your local demo/testing area into a published demo. Preact-CLI is generally for applications, not libraries, hence why you see a lot of complexity in the previous `preact.config.js`. Microbundle, in theory, should allow you the best of both worlds here: Preact-CLI for a testing/demo application, Microbundle for bundling your library for production. 

Thanks for your feedback though. There was a few things I knew I wanted to touch on just having used this template a bit more since I opened this PR, but I'll definitely add some new things in based on your comments here.

**Review by developit — APPROVED** [review]
I think we should merge this!

**Comment by developit** [comment]
Regarding microbundle bundling dependencies: it will never bundle `"dependencies"` from the package.json, because those are "modules that this module require()'s" - bundling them would mean everyone gets two copies.

There are some options:
1. Install dependencies that you want to be bundled inline into your widget as `"devDependencies"`. They're only used during development, because the published/consumed widget includes its own bundled copies.
2. Run `microbundle --external none`: this disables Microbundle's default behavior of inferring whether dependencies should be bundled or not.

**Comment by camsloanftc** [comment]
Thanks @developit - it totally makes sense when you put it this way.

I was still treating the package.json like it was for the entire preact app (including demo), hence keeping things in dependencies still, but treating it more like a node module primarily for the widget itself makes sense in that I would move things over to devDependencies or do the second option of `microbundle --external none`.

I did run into an issue when doing that. It's definitely out of scope for this PR though so I will raise the issue elsewhere.

**Comment by rschristian** [comment]
@developit 

One unfortunate item I'm just now noticing is that Preact-CLI will inject both the SW & template into `src/` when using `preact create` and there's no way around that as of now.

For the widgets I'd argue that separating the dev/demo aspect out of `src/` and into its own directory like `demo/` is much more clear, but if the CLI is going to inject application stuff into `src/` then we don't have much in the way of options, at least without maybe making the CLI widget-template aware. 

Happen to have any thoughts on the matter?

**Pull request merged**